src/_http/index.js

```diff
@@ -151,14 +151,54 @@ function isValidTraceparentHeader(traceparentHeader) {
 
 function parseTags(tags) {
   if (tags === undefined || tags == null || tags == '') return null
-  if (typeof tags != 'object') {
-    throw new Error('Tags must be provided as an object!')
-  }
+  validateTags(tags)
   return Object.entries(tags)
     .map(e => e.join('='))
     .join(',')
 }
 
+function validateTags(tags) {
+  const maxTagEntries = 25
+  if (typeof tags != 'object') {
+    throw new Error('Tags must be provided as an object!')
+  }
+  if (Object.entries(tags).length > maxTagEntries) {
+    throw new Error('Tags are limited to 25 entries')
+  }
+  validateTagKeys(Object.keys(tags))
+  validateTagValues(Object.values(tags))
+}
+
+function validateTagKeys(keys) {
```

Comment:
feels a bit aggressive to fail the request due to problems with tags.

I recall you thought about the tradeoffs between cleaning up the data vs blowing up - but could we discuss that again here?

Another option would be to force values to a string + trim up the input; and clean out tags that don't meet the spec.

I'm imaging a scenario where a customer has automation injecting tags - not sure they'd want to start failing their queries due to issues with the tag formats.

Reply:
TL;DR - Fail fast. If I'm doing something wrong, tell me, and tell me right away - because as my system gets bigger, it becomes more difficult to track down the problem.

Longer story is we ended up discussing this a bit on the DB side as well, where we have similar validation going into place. It boils down to whether you consider tags as part of the "request" (more specifically, the query request, and thus should throw 4XX) or if it's simply metadata that can be ok to fail (with the query returning 200). Whatever that decision, the client-side validation should match with the intent that we simply perform the validation as early as possible. When you consider the varied things that customers tend to use tags for - one of the most critical is often billing/analytics use cases. It's often not acceptable to let things go unnoticed there for long periods of time, and if you do, it becomes really painful to fix in retrospect (as most product devs know far too well...😒). I wouldn't want to just add this validation to a new use case/feature. However, in this case, we're introducing it, as well as these semantics, out of the gate. If you do something wonky, we'll let you know, and you can fix it. Perhaps, without knowing what you're using the tags for, and assuming that it's fairly orthogonal to your use case...then sure, we can silently fail/sanitize. However, if you assume that tagging becomes a mission-critical part of a customer's business, then it actually makes sense to fail immediately. They'll be able to fix the issue faster, and then move on with debugging the rest of their data flow.

As far as sanitizing the input - I think it's dangerous to start doing that on a customer request as becomes more difficult, as a consumer, to detect where things fail. If my key is too long, do you just blindly truncate it? Do you truncate the head, or tail? Is that configurable? One gap right now where we take the approach you're suggesting is on duplicate keys, which Javascript will just silently overwrite. I considered that an acceptable gap for most developers given that it's a language-level semantic, but I'd also prefer to throw an exception even for this case for the above reasons on how tags are typically used.

Reply:
so the database is going to throw 400s for bad tags?

Is the same true for bad trace ids?